### PR TITLE
feat: Add before_send filter to all plugins

### PR DESCRIPTION
**Pull request by jgraichen**
All plugins (e.g. rack, sidekiq) and the agent offer a `before_send:`
initializer argument now that takes a lambda to filter points before
writing. The lambda must return the (modified) points that will be
written. If nothing is returned, nothing will be written. This way,
points can be modified (e.g. tags added/removed) or filtered away.

Example:

    Telegraf::Rack.new(
      before_send: lambda do |point, request:, **|
        # Do not emit metrics for this secret path
        return if %r{^/secret}.match?(request.path)

        # Never collect controller details (make debugging hard)
        point.tags.delete(:controller)
        point
      end,
    )

See the individual plugins for which arguments are given to the lambda.

The agents before_send filter will always only received the a list of
raw data points as the write! method takes it too. No additional
arguments will be passed on.

For Rails, one can set before_send filter via the following config keys:

  * config.telegraf.before_send
  * config.telegraf.rack.before_send
  * config.telegraf.grape.before_send
  * config.telegraf.active_job.before_send
  * config.telegraf.sidekiq.before_send

### CODE REVIEW

File: lib/telegraf/railtie.rb [hunk 1]
```diff
@@ -48,48 +48,59 @@ class Railtie < ::Rails::Railtie
...
     config.telegraf.rack.tags = {}
+    config.telegraf.rack.before_send = nil
 
     # Install request instrumentation
     config.telegraf.instrumentation = true
 
     # Install Grape instrumentation
     config.telegraf.grape = ::ActiveSupport::OrderedOptions.new
     config.telegraf.grape.enabled = defined?(::Grape)
+    config.telegraf.grape.before_send = nil
```
Comment (franzliedke) [inline lib/telegraf/railtie.rb:66]: This one doesn't seem to be passed on. 🧐

File: spec/telegraf/agent_spec.rb [hunk 2]
```diff
@@ -16,9 +16,41 @@
...
+  describe '@before_send' do
+    let(:kwargs) { {before_send: before_send} }
+    let(:before_send) do
+      lambda {|data|
+        data.delete_if {|line| line[:tags].key?(:drop) }
+        data
+      }
+    end
+
+    it 'drops all points' do
```
Comment (franzliedke) [inline spec/telegraf/agent_spec.rb:34]:
```suggestion
    it 'drops all points when all points match' do
```

Is that what you mean?

File: spec/telegraf/rack_spec.rb [hunk 3]
```diff
@@ -101,4 +105,27 @@
...
+      let(:before_send) do
+        lambda {|point, request:, **|
+          return if %r{^/healthcheck}.match?(request.path)
+
+          point.values[:path] = request.path
+          point
+        }
+      end
+
+      it('drops matching points') do
```
Comment (franzliedke) [inline spec/telegraf/rack_spec.rb:122]: Nitpick... Hey, you're mixing up the DSL styles. 🤓 

```suggestion
      it 'drops matching points' do
```

Here and elsewhere. 🙈

**Review by franzliedke — APPROVED** [review]
Looks good!

The only relevant oversight I found is that `before_send` isn't used for the Grape plugin, I think.

File: lib/telegraf/railtie.rb (hunk 1 above)
Comment (jgraichen) [reply lib/telegraf/railtie.rb:66]: Correct. `grape` only adds data to the point in the rack env.

Will remove that one.

**Pull request merged**